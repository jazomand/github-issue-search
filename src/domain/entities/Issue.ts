import Label from './Label';

class Issue{
    constructor (
        public id: number,
        public title: string,
        public labels: Label[],
        public url: string,
        public repositoryUrl: string,
        public state: string,
        public createdAt: Date,
        public updatedAt: Date|null,
        public closedAt: Date|null,
        public body: string
    ) {}
};

export default Issue;