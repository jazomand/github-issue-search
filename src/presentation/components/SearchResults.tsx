import React, { useEffect, useState } from 'react';
import useKeyPress, {keys} from '../hooks/useKeyPress';
import Issue from '../../domain/entities/Issue';
import { Card, CardContent, CardHeader, Chip, Typography } from '@material-ui/core';

function SearchResults({issues, selectResult, setSelectResult} : {issues: Issue[], selectResult: boolean, setSelectResult: {(selectResult: boolean): void}}) {

    const [selectedIndex, setSelectedIndex] = useState(-1);
    const downKeyPress = useKeyPress(keys.arrowDown.keyCode);
    const upKeyPress = useKeyPress(keys.arrowUp.keyCode);
    const escKeyPress = useKeyPress(keys.escape.keyCode);

    useEffect(()=>{
        if (selectResult && downKeyPress){
            if (selectedIndex < issues.length - 1){
                document.getElementById(`card-${issues[selectedIndex + 1].id}`)?.scrollIntoView({ behavior: 'smooth', block: 'center' });
                setSelectedIndex(selectedIndex + 1);
            }
        }
    },[downKeyPress]);

    useEffect(()=>{
        if (selectResult && upKeyPress){
            setSelectedIndex(selectedIndex - 1);
            if (selectedIndex > 0){
                document.getElementById(`card-${issues[selectedIndex - 1].id}`)?.scrollIntoView({ behavior: 'smooth', block: 'center' });
            }else{
                setSelectResult(false);
            }
        }
    },[upKeyPress]);

    useEffect(()=>{
        if (selectResult && escKeyPress){
            setSelectResult(false);
            setSelectedIndex(-1);
        }
    },[escKeyPress]);

    useEffect(()=>{
        if (selectResult){
            setSelectedIndex(0);
        }else{
            setSelectedIndex(-1);
        }
    },[selectResult, issues]);

    if (issues.length > 100){
        issues = issues.slice(0,100);
    }
    return (<>
        {issues.map((issue, index) => 
            <Card 
                id={`card-${issue.id}`} 
                style={{marginTop: '2vh', padding: '1vh'}}
                key={issue.id} 
                raised={index === selectedIndex}
                onClick={()=>{
                    setSelectResult(true);
                    setSelectedIndex(index);
                }}>
                <CardHeader 
                    data-testid={`card-title-${issue.id}`}
                    title={issue.title}
                />
                <CardContent>
                    <Typography variant="body2" color="textSecondary" component="p">
                        {issue.body}
                    </Typography>
                </CardContent>
                <CardContent>
                    {issue.labels.map(label => 
                        <Chip key={label.id} label={label.name} style={{backgroundColor: `#${label.color}`, marginRight: '1vh'}} />
                    )}
                </CardContent>
            </Card>
        )}
    </>);
}

export default SearchResults;