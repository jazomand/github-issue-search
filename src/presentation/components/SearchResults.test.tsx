import React from 'react';
import { 
  render,
  screen
} from '@testing-library/react';
import SearchResults from './SearchResults';
import Issue from '../../domain/entities/Issue';

test('SearchResults renders correctly', () => {
    const mockIssues : Array<Issue> = [
        new Issue(
            1, 'ReallyBigIssue', [], 
            'https://issueurl.com', 'https://repositoryurl.com', 
            'open', new Date(), null, null, 'This is a reallyBigIssue'),
        new Issue(
            2, 'ReallySmallIssue', [], 
            'https://issueurl.com', 'https://repositoryurl.com', 
            'closed', new Date(), null, null, 'This is a reallySmallIssue'),
    ];
    const {asFragment} = render(<SearchResults issues={mockIssues} selectResult={false} setSelectResult={()=>{}} />);
    
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByTestId(`card-title-1`)).toHaveTextContent('ReallyBigIssue');
    expect(screen.getByTestId(`card-title-2`)).toHaveTextContent('ReallySmallIssue');
});