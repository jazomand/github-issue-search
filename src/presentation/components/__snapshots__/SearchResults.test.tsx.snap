// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    id="card-1"
    style="margin-top: 2vh; padding: 1vh;"
  >
    <div
      class="MuiCardHeader-root"
      data-testid="card-title-1"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          ReallyBigIssue
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
      >
        This is a reallyBigIssue
      </p>
    </div>
    <div
      class="MuiCardContent-root"
    />
  </div>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    id="card-2"
    style="margin-top: 2vh; padding: 1vh;"
  >
    <div
      class="MuiCardHeader-root"
      data-testid="card-title-2"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          ReallySmallIssue
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
      >
        This is a reallySmallIssue
      </p>
    </div>
    <div
      class="MuiCardContent-root"
    />
  </div>
</DocumentFragment>
`;
